core/src/main/java/edu/wpi/grip/core/sources/VideoFileSource.java

```diff
@@ -0,0 +1,194 @@
+package edu.wpi.grip.core.sources;
+
+import edu.wpi.grip.core.Source;
+import edu.wpi.grip.core.events.SourceHasPendingUpdateEvent;
+import edu.wpi.grip.core.events.SourceRemovedEvent;
+import edu.wpi.grip.core.sockets.OutputSocket;
+import edu.wpi.grip.core.sockets.SocketHint;
+import edu.wpi.grip.core.sockets.SocketHints;
+import edu.wpi.grip.core.util.ExceptionWitness;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.eventbus.EventBus;
+import com.google.common.eventbus.Subscribe;
+import com.google.inject.assistedinject.Assisted;
+import com.google.inject.assistedinject.AssistedInject;
+import com.thoughtworks.xstream.annotations.XStreamAlias;
+
+import org.bytedeco.javacpp.opencv_core.Mat;
+import org.bytedeco.javacv.FFmpegFrameGrabber;
+import org.bytedeco.javacv.Frame;
+import org.bytedeco.javacv.FrameGrabber;
+import org.bytedeco.javacv.OpenCVFrameConverter;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * A source for a video file input.
+ */
+@XStreamAlias("grip:VideoFile")
+public class VideoFileSource extends Source {
```

Comment:
There seems to be duplicate code between this and the Camera source. Can any of that be refactored into a base class?

Also, I notice you are using a thread. We had issues with threads and their weirdness. Any reason you didn't implement this as a restarting service?